Implement context and config for SnowthClient

dhaifley opened the pull request:
### This PR:

- Is completely backwards compatible with all previous versions of gosnowth. All changes introduced by this PR are non-breaking for any existing code using gosnowth.
- Introduces a Config type which allows creation of SnowthClient values with configurable properties. Values that can be set with configuration are:
  - DialTimeout sets a timeout value for the initial connection to an IRONdb server. Its default value is 500 milliseconds.
  - Discover sets whether or not the client should attempt to discover other IRONdb nodes using topology data retrieved from its provided servers. Its default value is false.
  - Servers sets the addresses of the IRONdb servers that should be used by the client.
  - Timeout sets a timeout value for the full duration of requests to IRONdb servers.  Its default value is 10 seconds.
  - WatchInterval sets the frequency at which the WatchAndUpdate() method checks for updates to the active status of the configured IRONdb nodes. Its default is 30 seconds.
- Implements context aware versions of all methods provided by SnowthClient.
  - These methods follow the standard library naming practice of appending 'Context' to the end of the name of the non-context aware method, and including a context.Context argument as the first parameter of the new method.
  - These methods all fully support context deadlines and cancellation, and the context value provided to them is attached to all requests sent to IRONdb and can be accessed in the functions provided via SetRequestFunc() using the r.Context() function.
- Includes unit tests for all new configuration and context functionality.
- Updates all dependencies to the newest versions.
- Adds account_id and check_name values to data written with SnowthClient.WriteHistogram().

bdunavant commented on config.go:
```diff
@@ -0,0 +1,186 @@
+func (c *Config) WithServers(servers ...string) *Config {
+	c.Servers = []string{}
+	for _, svr := range servers {
+		if _, err := url.Parse(svr); err == nil {
```
If they don't parse, do we care about any error messages?  We do this in UnmarshalJSON above too.

dhaifley replied:
Maybe, but also maybe not. These With* functions would be used when initially building a new Config value, and are intended to be chained together. So, all of them follow a pattern of just not assigning the value passed if it does not pass validation. If we were to having them also return an error, that would prevent being able to chain them together. Also, it is important to keep in mind, that these methods are just used to build a configuration value that is then used to initialize a SnowthClient value, not actually setup the client itself. When the SnowthClient is initialized is when the checks happen that do more than just validate if the address formats are valid. Snowth.NewClient() checks that at least one of the provided servers can be connected, and fails with an error if no servers can be connected. So, if no valid server addresses are provided when setting up a Config value, an error will occur when an attempt is made to use it when initializing a SnowthClient value.

dhaifley commented:
Thanks for the feedback.  I have added guards and some validation to the assignment of config values with the With* methods and to the JSON unmarshaling process.  I implemented the custom JSON marshaling interface for Config values using a map[string]interface{} intermediary because I wanted to allow the timeout and interval Config fields to be the standard library time.Duration type, as this is how they are used when setting up a SnowthClient.  And, while time.Duration can unmarshal from JSON with only a struct tag, it only does so if the input is an int representing nanoseconds, which I think is a weird expectation for the JSON representation.  This way allows the JSON to contain the standard string representations for time.Duration, like "100ms", "1s", "24h", etc.  I could have created a custom duration type implementing JSON the marshaling interface for these values, or a custom intermediary struct type for marshaling Config, but I figured this way was pretty much equivalent.

maier commented:
Yeah, that's a give and take scenario. For a config that can/will be stored on disk and parsed, I usually don't use the native types - just use string for everything and during the load use the appropriate parsers to vet the values set. For a progmatic config passed into a method I can see using the native types as that just makes more sense; vs having a user set strings only for them to be parsed to durations which the user could have set in the first place :)

dhaifley commented:
Yeah, the JSON support for Config was implemented more as a possible convenience for someone in the future.  Obviously there are no existing use cases for this, and the intended purpose of the Config struct is the latter you mentioned: a place to collect all the values used to initialize a SnowthClient value and pass to NewClient(), that is a programmatic configuration structure.  There are no existing JSON configurations for gosnowth anywhere on disk, and in the downstream use of this that I am working on (GoAPI) it would not be used that way.
